### PR TITLE
#24: fix parallel requests for RefreshTokenFeature

### LINKED ISSUE #24: RefreshTokenFeature try to refresh token twice for parallel requests
If we made two requests with invalid token, first of them will refresh correctly, but second will try to refresh with old refresh token and fire unauth action.

have this solution:

            install(RefreshTokenFeature) {
                updateTokenHandler = {
                    // Локально сохраняем текущий токен авторизации
                    val currentAuthToken = keyValueStorage.token

                    // Блокируем выполнение корутины через мьютекс, чтобы одновременно не
                    // отправлялось несколько запросов на обновление токена.
                    refreshTokenHttpFeatureMutex.lock()

                    // Если локальный токен авторизации currentAuthToken после захода в
                    // критическую секцию уже не равен токену в хранилище (был изменен в результате
                    // другого запроса), то посылать еще запрос на обновление токена скорее всего
                    // не надо, поэтому пытаемся выполнить предыдущий запрос снова.
                    if (currentAuthToken != keyValueStorage.token) {
                        refreshTokenHttpFeatureMutex.unlock()
                        true
                    }
                    // Если же локальный токен равен токену в хранилище, то значит его надо обновить,
                    // отправляем запрос на обновление токена.
                    else {
                        try {
                            authRepository.updateToken()
                            true
                        } catch (error: Exception) {
                            authRepository.clearUser()
                            profileRepository.clearProfile()
                            false
                        } finally {
                            refreshTokenHttpFeatureMutex.unlock()
                        }
                    }
                }
            }

### PR DESCRIPTION
**Pull request by Tetraquark**
Closes #24

### CODE REVIEW

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt [hunk 1]
```diff
@@ -13,35 +13,64 @@ import io.ktor.client.statement.HttpReceivePipeline
...
         override fun prepare(block: Config.() -> Unit) = Config().apply(block).build()
 
         override fun install(feature: RefreshTokenFeature, scope: HttpClient) {
             scope.receivePipeline.intercept(HttpReceivePipeline.After) {
-                if (context.response.status == HttpStatusCode.Unauthorized && feature.updateTokenHandler()) {
-                    val requestBuilder = HttpRequestBuilder().takeFrom(context.request)
-                    val result: HttpResponse = context.client.request(requestBuilder)
-                    proceedWith(result)
+                if (context.response.status == HttpStatusCode.Unauthorized) {
+                    val currentToken = feature.getCurrentToken()
```
Comment (Alex009) [inline network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt:44]: here should be reading token of this request

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt [hunk 2]
```diff
@@ -13,35 +13,64 @@ import io.ktor.client.statement.HttpReceivePipeline
...
+
+                    refreshTokenHttpFeatureMutex.lock()
+
+                    // If the local current token is already changed, then the feature will
+                    // repeat the previous request
+                    if (currentToken != feature.getCurrentToken()) {
```
Comment (Alex009) [inline network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt:50]: `if(!feature.isTokenActual(context.request))`

```
feature.isTokenActual = { request ->
    request.header["Authorization"] == keyValueStorage.token
}
```

Comment (Alex009) [reply network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt:50]: `isCredentialsActual`

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt [hunk 3]
```diff
@@ -13,35 +13,64 @@ import io.ktor.client.statement.HttpReceivePipeline
...
     companion object Feature : HttpClientFeature<Config, RefreshTokenFeature> {
 
+        private val refreshTokenHttpFeatureMutex = Mutex()
+
         override val key = AttributeKey<RefreshTokenFeature>("RefreshTokenFeature")
 
         override fun prepare(block: Config.() -> Unit) = Config().apply(block).build()
 
         override fun install(feature: RefreshTokenFeature, scope: HttpClient) {
             scope.receivePipeline.intercept(HttpReceivePipeline.After) {
```
Comment (Alex009) [inline network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt:42]: named argument

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt [hunk 4]
```diff
@@ -13,35 +13,64 @@ import io.ktor.client.statement.HttpReceivePipeline
...
-                if (context.response.status == HttpStatusCode.Unauthorized && feature.updateTokenHandler()) {
-                    val requestBuilder = HttpRequestBuilder().takeFrom(context.request)
-                    val result: HttpResponse = context.client.request(requestBuilder)
-                    proceedWith(result)
+                if (context.response.status == HttpStatusCode.Unauthorized) {
```
Comment (Alex009) [inline network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt:43]: fast return
```
if (context.response.status != HttpStatusCode.Unauthorized) {
    proceedWith(it)
    return
}
```

Comment (Alex009) [reply network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt:43]: and other ifs

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt (hunk 2 above)
Comment (Tetraquark) [reply network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt]: Fixed.

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt (hunk 3 above)
Comment (Tetraquark) [reply network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt]: Fixed.

File: network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt (hunk 4 above)
Comment (Tetraquark) [reply network/src/commonMain/kotlin/dev/icerock/moko/network/features/RefreshTokenFeature.kt]: Fixed.

**Pull request merged**